pyccl/halos/massdef.py

```diff
@@ -248,49 +218,130 @@ def from_name(cls, name):
         Delta, rho_type = name[:-1], parser[name[-1]]
-        return lambda cm=None: cls(Delta, rho_type, concentration=cm)  # noqa
+        return lambda: cls(Delta, rho_type)  # noqa
 
+    @classmethod
+    def from_specs(cls, mass_def=None, *,
+                   mass_function=None, halo_bias=None, concentration=None):
+        """Instantiate mass definition and halo model ingredients.
+
+        Unspecified halo model ingredients are ignored. ``mass_def`` is always
+        instantiated.
+
+        Parameters
+        ----------
+        mass_def : MassDef, str or None, optional
+            Mass definition. If a string, instantiate from its name. If None,
+            obtain the one from the first specified halo model ingredient.
+            The default is None.
+        mass_function, halo_bias, concentration : \
+            (MassFunc, HaloBias, Concentration), str or None, optional
+            Halo model ingredients. Strings are auto-instantiated using
+            ``mass_def``. None values are ignored. The defaults are None.
+
+        Returns
+        -------
+        mass_def : MassDef
+
+        mass_function : MassFunction, if specified
+
+        halo_bias : HaloBias, if specified
+
+        concentration : Concentration, if specified
+
+        Raises
+        ------
+        ValueError
+            If mass definition cannot be retrieved from halo model ingredients.
+        ValueError
+            If mass definitions are inconsistent.
+        """
+        values = mass_function, halo_bias, concentration
+        idx = [value is not None for value in values]
+
+        # Filter only the specified ones.
+        values = np.array(values)[idx]
+        names = np.array(["mass_function", "halo_bias", "concentration"])[idx]
+        Types = np.array([MassFunc, HaloBias, Concentration])[idx]
+
+        # Sanity check.
+        if mass_def is None:
+            for name, value in zip(names, values):
+                if isinstance(value, str):
+                    raise ValueError(f"Need mass_def if {name} is str.")
+
+        # Instantiate mass_def.
+        if mass_def is not None:
+            mass_def = cls.create_instance(mass_def)  # instantiate directly
+        else:
+            mass_def = values[0].mass_def  # use the one in HMIngredients
 
-@warn_api(pairs=[('c_m', 'concentration')])
-def MassDef200m(concentration='Duffy08'):
-    r""":math:`\Delta = 200m` mass definition.
+        # Instantiate halo model ingredients.
+        out = []
+        for name, value, Type in zip(names, values, Types):
+            instance = Type.create_instance(value, mass_def=mass_def)
+            out.append(instance)
 
-    Args:
-        concentration (string): concentration-mass relation.
-    """
-    return MassDef(200, 'matter', concentration=concentration)
+        # Check mass definition consistency.
+        if hmd := set([x.mass_def for x in out]):
+            if hmd != set([mass_def]):
```

Comment:
This seems very obscure. Why not `if set([x.mass_def for x in out]) != set([mass_def]):`

Reply:
I know the operator, I'm just saying this is rather obscure. My suggestion would check both things in one go with no added overhead.

Reply:
That's what I'm saying, it can't be done in one go.

Reply:
why not? hmd is not used anywhere else, and if the set is empty, then it won't be equal to `set([mass_def])`

Reply:
Because if no halo model ingredients have been passed, doing what you proposed will raise an error. You have to check first, if `out` is empty first. I replaced it with a single-liner, check if you like it now.

Reply:
I don't think you need the first bit of that one-liner, but that's fine, I can live with it.

Reply:
I literally tried it and got an error, the first bit does all the work.